### PR TITLE
[SYCL-MLIR][DetectReduction] Disallow transformation when exists may alias operations in loop

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Original loop:
```
for (i) 
  A[0] += B[i];
```
Optimized loop:
```
a = A[0];
for (i)
  a += B[i];
A[0] = a;
```
This is incorrect if A and B may alias because `B[i]` may be different after updating `A[0]`.

### CODE REVIEW

**Comment by whitneywhtsang** [comment]
Plan to merge this PR after implementing a fix. But please still review it.

File: polygeist/test/polygeist-opt/red.mlir
Comment (sommerlukas) [inline polygeist/test/polygeist-opt/red.mlir:144]: Nit: Add a comment specifying which operations form the may-alias.

File: polygeist/tools/cgeist/Test/Verification/redstore.c
Comment (sommerlukas) [inline polygeist/tools/cgeist/Test/Verification/redstore.c:3]: Do I understand correctly that we currently need SYCL compilation to understand `__restrict__`? 

Do you plan to fix that in a different PR before merging this one?

Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/redstore.c:3]: > Do I understand correctly that we currently need SYCL compilation to understand __restrict__?

Yes, you understand it correctly, we only add attributes to SYCL functions. One reason is to avoid fixing all existing non SYCL test cases. 

> Do you plan to fix that in a different PR before merging this one?

To avoid fixing all existing non SYCL test cases, actually I don't plan to fix that, I can look into adding an option to allow adding attributes for non SYCL functions.

File: polygeist/test/polygeist-opt/red.mlir
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/red.mlir:144]: done.

File: polygeist/tools/cgeist/Test/Verification/redstore.c
Comment (etiotto) [reply polygeist/tools/cgeist/Test/Verification/redstore.c:3]: Eventually we will need to process host code in MLIR so this limitation will need to be addressed but for now we can side step the problem.

File: polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir
Comment (etiotto) [reply polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir:2]: Yes I think we should check whether all incoming accessor (in the GPU kernel) may overlap with a runtime check. IF the check succeeds invoke the function containing the kernel workload (the lambda body) where all accessors have an attribute indicating their buffers do not overlap. Conversely, if the check fails, call a clone of that function.

File: polygeist/lib/Dialect/Polygeist/Transforms/DetectReduction.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/DetectReduction.cpp:397]: we need to return on all path, or else there is warnings.

File: polygeist/tools/cgeist/Test/polybench/linear-algebra/kernels/2mm/2mm.c
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/polybench/linear-algebra/kernels/2mm/2mm.c:18]: done, change the CHECKs.

File: polygeist/lib/Dialect/Polygeist/Transforms/DetectReduction.cpp
Comment (whitneywhtsang) [reply polygeist/lib/Dialect/Polygeist/Transforms/DetectReduction.cpp:375]: done.

**Review by etiotto — COMMENTED** [review]
Changes from commit 3f4755e looks fine to me.

File: polygeist/tools/cgeist/Test/Verification/redstore.c
Comment (whitneywhtsang) [reply polygeist/tools/cgeist/Test/Verification/redstore.c]: added option control.

**Review by sommerlukas — APPROVED** [review]

File: polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/sycl/matrix_multiply_reduction.mlir]: Resolved.

**Review by etiotto — APPROVED** [review]

**Pull request merged**
